### PR TITLE
thread exit cleanup

### PR DESCRIPTION
**Pull request by wjhun**
This contains a variety of clean up around unix land to allow us to free up structures on thread exit. Mostly this consists of straightening out refcounts in various structures (facilitated by adding a refcount interface - a small improvement over fetch_and_add) and finishing cleanup for various types of objects (particularly with poll waiting). Note that yet more additions have been made to closures - specifically the ability to embed a closure struct within another struct, which I used for the various refcount completions.

See #456

### CODE REVIEW

File: src/unix/blockq.c [hunk 1]
```diff
@@ -227,6 +228,7 @@ sysreturn blockq_check_timeout(blockq bq, thread t, blockq_action a,
 
     bi->a = a;
     bi->t = t;
+    refcount_reserve(&t->refcount);
```
Comment (briankoco) [inline src/unix/blockq.c]: Should probably be released if the call to register_timer() fails a couple of lines below

File: src/unix/poll.c [hunk 2]
```diff
@@ -356,9 +399,14 @@ static epoll_blocked alloc_epoll_blocked(epoll e)
     epoll_blocked w = unix_cache_alloc(get_unix_heaps(), epoll_blocked);
     if (w == INVALID_ADDRESS)
 	return w;
-    w->refcnt = 1;
+    epoll_debug("w %p\n", w);
+
+    /* initial reservation released on thread wakeup (or direct return) */
+    init_refcount(&w->refcount, fill_closure(&w->free, epoll_blocked_free, w));
     w->t = current;
```
Comment (briankoco) [inline src/unix/poll.c]: This should be taking a ref on current, right?

File: src/unix/blockq.c (hunk 1 above)
Comment (wjhun) [reply src/unix/blockq.c]: Thanks; I added a check for register_timer failure.

File: src/unix/poll.c [hunk 3]
```diff
@@ -6,14 +6,18 @@
 #define epoll_debug(x, ...)
```
[the remark this answers is not in the document]
Comment (wjhun) [reply src/unix/poll.c]: Yes, I intend to add them here and in a few other places in a follow-up PR.

**Pull request merged**